File: MaximalSquare.py
"""
Given a 2D binary matrix filled with 0's and 1's, find the largest square containing all 1's and return its area.

For example, given the following matrix:

1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0
Return 4.
"""


class Solution:
    # @param {character[][]} matrix
    # @return {integer}
    def maximalSquare(self, matrix):
        if not matrix or not matrix[0]:
            return 0
        m, n = len(matrix), len(matrix[0])
        max_square = [[0] * n for _ in xrange(m)]
        for i in xrange(m):
            max_square[i][0] = int(matrix[i][0])
        for j in xrange(n):
            max_square[0][j] = int(matrix[0][j])

        for i in xrange(1, m):
            for j in xrange(1, n):
                if matrix[i][j] == '1':
                    max_square[i][j] = min(max_square[i-1][j-1], max_square[i][j-1], max_square[i-1][j]) + 1

        return max([max(row) for row in max_square]) ** 2

File: Leetcode-Java/MaximalSquare.java
/**
 * Created by yhf on 15/6/13.
 */

/*
Given a 2D binary matrix filled with 0's and 1's, find the largest square containing all 1's and return its area.

For example, given the following matrix:

1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0

Return 4.

dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1]) + 1
*/

public class MaximalSquare {
    public static int maximalSquare(char[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }
        int m = matrix.length;
        int n = matrix[0].length;
        int[][] dp = new int[m][n];

        for (int j = 0; j < n; j++) {
            dp[0][j] = matrix[0][j] - '0';
        }
        for (int i = 0; i < m; i++) {
            dp[i][0] = matrix[i][0] - '0';
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][j] == '1') {
                    dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]), dp[i-1][j-1]) + 1;
                } else {
                    dp[i][j] = 0;
                }
            }
        }

        int max = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                max = Math.max(dp[i][j], max);
            }
        }

        return max * max;
    }

    public static void main(String[] args) {
        /*char[][] matrix = {
                {'1', '0', '1', '0', '0'},
                {'1', '0', '1', '1', '1'},
                {'1', '1', '1', '1', '1'},
                {'1', '0', '0', '1', '0'}};*/
        char[][] matrix = {
                {'1', '1'},
                {'1', '1'}};
        int num = maximalSquare(matrix);
    }
}
